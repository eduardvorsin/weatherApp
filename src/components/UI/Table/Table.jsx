import PropTypes from 'prop-types';
import classes from './Table.module.css';

export default function Table({
  className,
  headerCells,
  dataRows,
  renderInDataCell,
  testId,
}) {
  return (
    <div
      className={`${classes['table-overlay']} ${className}`}
      data-testid={testId}
    >
      <div className={classes['table-container']}>
        <table
          className={classes.table}
        >

          <thead className={classes.table__head}>
            <tr className={classes.table__row}>
              {headerCells.map((cell) => (
                <th
                  className={classes['table__head-cell']}
                  key={cell.name}
                >
                  {cell?.name}
                </th>
              ))}
            </tr>
          </thead>

          <tbody className={classes.table__body}>
            {dataRows.map((row) => (
              <tr key={row.id} className={classes.table__row}>
                {
                  Object.entries(row.data).map(([key, value]) => (
                    <td
                      className={classes['table__data-cell']}
                      key={key}
                    >
                      {
                        renderInDataCell
                          ? renderInDataCell(key, value)
                          : value
                      }
                    </td>
                  ))
                }
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
}

Table.defaultProps = {
  className: '',
  headerCells: [],
  dataRows: [],
  renderInDataCell: null,
  testId: '',
};

Table.propTypes = {
  className: PropTypes.string,
  // eslint-disable-next-line react/forbid-prop-types
  headerCells: PropTypes.arrayOf(PropTypes.object),
  // eslint-disable-next-line react/forbid-prop-types
  dataRows: PropTypes.arrayOf(PropTypes.object),
  renderInDataCell: PropTypes.func,
  testId: PropTypes.string,
};
