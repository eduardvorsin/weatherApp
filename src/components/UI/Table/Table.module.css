.table-overlay {
  position: relative;
}

.table-overlay::before,
.table-overlay::after {
  position: absolute;
  top: 0;
  content: '';
  height: 100%;
  width: 20px;
}

.table-overlay::after {
  background: linear-gradient(90deg, rgba(255, 255, 255, 0) 0%, rgba(255, 255, 255, 0.2) 100%);
  right: 0;
}

.table-overlay::before {
  background: linear-gradient(90deg, rgba(255, 255, 255, 0.2) 0%, rgba(255, 255, 255, 0) 100%);
  left: 0;
}

@media(min-width:1100px) {
  .table-overlay {
    position: static;
  }

  .table-overlay::before,
  .table-overlay::after {
    display: none;
  }
}

.table-container {
  overflow-x: auto;
}

.table {
  border-collapse: collapse;
  border-spacing: 0px;
  text-align: center;
  width: 100%;
}

.table__head {}

.table__row {}

.table__body {}

.table,
.table__data-cell,
.table__head-cell {
  border: 1px solid var(--brand-color);

}

.table__data-cell,
.table__head-cell {
  padding: 10px 5px;
}
