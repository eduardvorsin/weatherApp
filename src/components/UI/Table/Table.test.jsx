import { render, screen } from '@testing-library/react';

import Table from './Table';

describe('Table component tests', () => {
  it('renders correctly', () => {
    const headerCells = [
      { name: '1' },
      { name: '2' },
      { name: '3' },
    ];

    const dataRows = [{
      id: 0,
      data: [1, 2, 3],
    }];

    render(<Table
      className="test-table"
      headerCells={headerCells}
      dataRows={dataRows}
      renderInDataCell={null}
    />);

    expect(screen.getByRole('table')).toBeInTheDocument();
    expect(screen.getAllByRole('columnheader')).toHaveLength(3);
    expect(screen.getAllByRole('cell')).toHaveLength(3);
  });

  it('renders correctly without props', () => {
    render(<Table />);

    expect(screen.getByRole('table')).toBeInTheDocument();
    expect(screen.queryByRole('columnheader')).not.toBeInTheDocument();
    expect(screen.getAllByRole('row')).toHaveLength(1);
    expect(screen.queryByRole('cell')).not.toBeInTheDocument();
  });

  it('correctly renders the div passed to renderInItem', () => {
    const headerCells = [
      { name: 'a' },
      { name: 'b' },
      { name: 'c' },
      { name: 'd' },
    ];

    const dataRows = [{
      id: 0,
      data: {
        a: 'red',
        b: 'orange',
        c: 'blue',
        d: 'purple',
      },
    }];

    render(<Table
      headerCells={headerCells}
      dataRows={dataRows}
      renderInDataCell={(_, value) => (
        <div data-testid="test-div">{value}</div>
      )}
    />);

    expect(screen.getAllByRole('cell')).toHaveLength(4);
    expect(screen.getAllByTestId('test-div')).toHaveLength(4);
    expect(screen.getByText(headerCells[0].name)).toBeInTheDocument();
    expect(screen.getByText(dataRows[0].data.c)).toBeInTheDocument();
  });

  it('if the renderInItem prop is not passed, then just render <td></td> with the values', () => {
    const headerCells = [
      { name: 'a' },
      { name: 'b' },
      { name: 'c' },
    ];

    const dataRows = [{
      id: 0,
      data: ['red', 'orange', 'blue'],
    }];

    render(<Table
      headerCells={headerCells}
      dataRows={dataRows}
    />);

    expect(screen.getAllByRole('cell')).toHaveLength(3);
    expect(screen.getByText(headerCells[2].name)).toBeInTheDocument();
    expect(screen.getByText(dataRows[0].data[1])).toBeInTheDocument();
  });

  it('the classname is correctly assigned to the element', () => {
    render(<Table
      className="test-class"
      testId="table"
    />);

    expect(screen.getByTestId('table')).toHaveClass('test-class');
  });

  it('snapshot of an empty table', () => {
    render(<Table testId="table" />);

    expect(screen.getByRole('table')).toMatchSnapshot();
  });

  it('snapshot with mock data', () => {
    const headerCells = [
      { name: 'animal1' },
      { name: 'animal2' },
      { name: 'animal3' },
    ];

    const dataRows = [{
      id: 0,
      data: ['cat', 'dog', 'parrot'],
    }];

    render(<Table
      headerCells={headerCells}
      dataRows={dataRows}
      testId="table"
    />);

    expect(screen.getByRole('table')).toMatchSnapshot();
  });
});
