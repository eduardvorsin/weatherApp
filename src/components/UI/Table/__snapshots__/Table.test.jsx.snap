// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component tests snapshot of an empty table 1`] = `
<table
  class="table"
>
  <thead
    class="table__head"
  >
    <tr
      class="table__row"
    />
  </thead>
  <tbody
    class="table__body"
  />
</table>
`;

exports[`Table component tests snapshot with mock data 1`] = `
<table
  class="table"
>
  <thead
    class="table__head"
  >
    <tr
      class="table__row"
    >
      <th
        class="table__head-cell"
      >
        animal1
      </th>
      <th
        class="table__head-cell"
      >
        animal2
      </th>
      <th
        class="table__head-cell"
      >
        animal3
      </th>
    </tr>
  </thead>
  <tbody
    class="table__body"
  >
    <tr
      class="table__row"
    >
      <td
        class="table__data-cell"
      >
        cat
      </td>
      <td
        class="table__data-cell"
      >
        dog
      </td>
      <td
        class="table__data-cell"
      >
        parrot
      </td>
    </tr>
  </tbody>
</table>
`;
